src/coordinator/ts/m3db/convert.go

```diff
@@ -0,0 +1,170 @@
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3db
+
+import (
+	"time"
+
+	"github.com/m3db/m3db/src/dbnode/encoding"
+	"github.com/m3db/m3db/src/dbnode/x/xio"
+	"github.com/m3db/m3x/ident"
+)
+
+const (
+	// todo(braskin): replace this with actual size if we can
+	blockReplicaLen = 10
+)
+
+// blockReplica contains the replicas for a single m3db block
+type blockReplica struct {
+	start     time.Time
+	blockSize time.Duration
+	replicas  []encoding.MultiReaderIterator
+}
+
+// ConvertM3DBSeriesIterators converts m3db SeriesIterators to SeriesBlocks
+// which are used to construct Blocks for query processing.
+func ConvertM3DBSeriesIterators(iterators encoding.SeriesIterators, iterAlloc encoding.ReaderIteratorAllocate) ([]SeriesBlocks, error) {
+	defer iterators.Close()
+	multiSeriesBlocks := make([]SeriesBlocks, iterators.Len())
+
+	for i, seriesIterator := range iterators.Iters() {
+		blockReplicas, err := blockReplicasFromSeriesIterator(seriesIterator, iterAlloc)
+		if err != nil {
+			return nil, err
+		}
+
+		series, err := seriesBlocksFromBlockReplicas(blockReplicas, seriesIterator)
+		if err != nil {
+			return nil, err
+		}
+
+		multiSeriesBlocks[i] = series
+	}
+
+	return multiSeriesBlocks, nil
+}
+
+func blockReplicasFromSeriesIterator(seriesIterator encoding.SeriesIterator, iterAlloc encoding.ReaderIteratorAllocate) ([]blockReplica, error) {
+	blockReplicas := make([]blockReplica, 0, blockReplicaLen)
+	for _, replica := range seriesIterator.Replicas() {
+		perBlockSliceReaders := replica.Readers()
+		for next := true; next; next = perBlockSliceReaders.Next() {
+			l, start, bs := perBlockSliceReaders.CurrentReaders()
+			readers := make([]xio.SegmentReader, l)
+			for i := 0; i < l; i++ {
+				reader := perBlockSliceReaders.CurrentReaderAt(i)
+				// NB(braskin): important to clone the reader as we need its position reset before
+				// we use the contents of it again
+				clonedReader, err := reader.Clone()
```

Comment:
how do we test this condition ?

Reply:
You could read the initial seriesIterator passed in; if you get values out of it, the clone is successful